File: day01.py
from typing import Tuple, List, TypeVar, Iterable, Callable

import math
import itertools


def read_input() -> List[int]:
    with open('day01.txt', 'r') as f:
        return list(map(int, f.readlines()))

def find_subset(data: Iterable[Tuple[int]], f: Callable[[Iterable[int]], int], equals: int) -> Tuple[int]:
    return next((g for g in data if f(g) == equals))

def part1() -> int:
    input_data = read_input()
    subset_2020 = find_subset(itertools.combinations(input_data, 2), sum, 2020)
    return math.prod(subset_2020)

def part2() -> int:
    input_data = read_input()
    subset_2020 = find_subset(itertools.combinations(input_data, 3), sum, 2020)
    return math.prod(subset_2020)

if __name__ == "__main__":
    print('Part1:', part1())
    print('Part2:', part2())

File: 2021/day01.hs
import Data.List

readInput :: IO [Int]
readInput = do
    content <- readFile "day01.txt"
    return $ map read $ lines content


groupsOf2 :: Num a => [a] -> [[a]]
groupsOf2 xs = [[x,y] | x:ys <- tails xs, y <- ys]

groupsOf3 :: Num a => [a] -> [[a]]
groupsOf3 xs = [[x,y,z] | x:ys <- tails xs, y:zs <- tails ys, z <- zs]

findSubsetEquals :: (Eq a, Num a) => ([a]->a) -> [[a]] -> a -> [a]
findSubsetEquals f (x:xs) n
            | f x == n = x
            | otherwise  = findSubsetEquals f xs n

part1 :: IO Int
part1 = do
    inputData <- readInput
    return $ product $ findSubsetEquals sum (groupsOf2 inputData) 2020

part2 :: IO Int
part2 = do
    inputData <- readInput
    return $ product $ findSubsetEquals sum (groupsOf3 inputData) 2020

